### PR TITLE
fix: shares for amount calculation should use freeFunds

### LINKED ISSUE #390: Question about recalculation of shares in withdraw
Should the recalculation of shares to burn on withdraw use `freeFunds` instead of `totalAssets`?

First  calculation of `shares` to `value` uses `freeFunds`
https://github.com/yearn/yearn-vaults/blob/master/contracts/Vault.vy#L1048

Second calculation of `value + totalLoss` to `shares` uses `totalAssets`

https://github.com/yearn/yearn-vaults/blob/master/contracts/Vault.vy#L1101

Share holder loses less share when second calculation is performed with `totalAssets`

### PR DESCRIPTION
**Pull request by steffenix**
fixes #390

### CODE REVIEW

File: contracts/Vault.vy [hunk 1]
```diff
@@ -813,6 +813,11 @@ def _calculateLockedProfit() -> uint256:
     else:
         return 0
 
+@view
+@internal
+def freeFunds() -> uint256:
```
Comment (jmonteer) [inline contracts/Vault.vy:818]: minor: I would use `_freeFunds` to be consistent with how the rest of internal view functions have been named

**Review by jmonteer — COMMENTED** [review]
added a comment for naming

otherwise LGTM

File: contracts/Vault.vy [hunk 2]
```diff
@@ -825,8 +830,7 @@ def _issueSharesForAmount(to: address, amount: uint256) -> uint256:
     if totalSupply > 0:
         # Mint amount of shares based on what the Vault is managing overall
         # NOTE: if sqrt(token.totalSupply()) > 1e39, this could potentially revert
-        freeFunds: uint256 = self._totalAssets() - self._calculateLockedProfit()
-        shares =  amount * totalSupply / freeFunds  # dev: no free funds
+        shares =  amount * totalSupply / self.freeFunds()  # dev: no free funds
```
Comment (fubuloubu) [inline contracts/Vault.vy:833]:
```suggestion
        shares =  amount * totalSupply / self._freeFunds()  # dev: no free funds
```

File: contracts/Vault.vy [hunk 3]
```diff
@@ -913,11 +917,10 @@ def _shareValue(shares: uint256) -> uint256:
 
     # Determines the current value of `shares`.
     # NOTE: if sqrt(Vault.totalAssets()) >>> 1e39, this could potentially revert
-    freeFunds: uint256 = self._totalAssets() - self._calculateLockedProfit()
 
     return (
         shares
-        * freeFunds
+        * self.freeFunds()
```
Comment (fubuloubu) [inline contracts/Vault.vy:923]:
```suggestion
        * self._freeFunds()
```

File: contracts/Vault.vy [hunk 4]
```diff
@@ -932,7 +935,7 @@ def _sharesForAmount(amount: uint256) -> uint256:
         return  (
             amount
             * self.totalSupply
-            / self._totalAssets()
+            / self.freeFunds()
```
Comment (fubuloubu) [inline contracts/Vault.vy:938]:
```suggestion
            / self._freeFunds()
```

Comment (fubuloubu) [inline contracts/Vault.vy:938]: Why this change?

Comment (steffenix) [reply contracts/Vault.vy:938]: Because the shares for the amount in the event of a loss should be calculated based on the free funds and not the totalAssets https://github.com/yearn/yearn-vaults/issues/390

File: contracts/Vault.vy (hunk 2 above)
Comment (steffenix) [reply contracts/Vault.vy]: Thanks for pointing this out made a change to the function name but not the call 🤦‍♂️

File: contracts/Vault.vy (hunk 4 above)
Comment (fubuloubu) [reply contracts/Vault.vy]: Okay. Should probably also adjust the `if` statement above (and cache that value too)

Comment (steffenix) [reply contracts/Vault.vy]: 👌

**Review by fubuloubu — APPROVED** [review]

**Comment by fubuloubu** [comment]
Should a test be added to cacth this behavior?

**Comment by steffenix** [comment]
> Should a test be added to cacth this behavior?

Should we add a function that exposes `_sharesForAmount` similar to `_shareValue` that's exposed by `pricePerShare`

**Comment by fubuloubu** [comment]
> > Should a test be added to cacth this behavior?
> 
> Should we add a function that exposes `_sharesForAmount` similar to `_shareValue` that's exposed by `pricePerShare`
> 
> Otherwise, the test will need to be performed on a withdraw
> 

On withdrawal sounds appropriate

**Review by jmonteer — APPROVED** [review]
